Compat: Disable wpautop for block-based posts

aduth opened the pull request:
Closes #2736

This pull request seeks to disable wpautop for Gutenberg posts. The markup structure of a block is defined as the result of the block implementation's `save` function, and does not need (and should not have) formatting of automatic paragraph application from `wpautop`. The changes here disable `the_content`'s filter of `wpautop` for posts containing blocks.

The challenge here is in managing posts which were originally authored in the Classic editor, then later in Gutenberg to add new content blocks (which will be a common occurrence after Gutenberg is merged). 

(See revised behavior: https://github.com/WordPress/gutenberg/pull/2806#issuecomment-332962480)

~To preserve paragraphs which had previously relied on the behavior of `autop`, the Gutenberg editor will apply `autop` during the parse step. This has the result of setting `<p>` into saved content, instead of relying on this to be applied later during `the_content`.~

~This is unlike some of the considerations proposed at #2708, where autop would be applied on a per-block basis in `the_content`. I am very wary of the performance impact of detecting and individually formatting blocks in content during a front-end filter. In the suggested use-case of an external editor, autop behavior would continue to apply for posts which do not contain blocks. It should never be the case that a post contains both blocks and legacy freeform content without explicit `<p>` tags if edited in Gutenberg, but it would be true that such an external client would need to pass new saved content in a block post with explicit tags as well. This is not complete compatibility, and it may be worth exploring whether `wpautop` should _also_ be applied server-side when saving a post containing both blocks and freeform content to capture these instances.~

__Testing instructions:__

1. Create a new post in the classic editor with a single paragraph of text
2. Save the post as a draft
3. Return to the posts list
4. Edit the post saved in Step 2 in the Gutenberg editor (hover in list, then click Gutenberg)
5. Add a new paragraph block with some text
6. Press Preview
7. Inspect content of preview
8. Note that both paragraphs are wrapped in `<p>` tags

aduth committed:
Compat: Disable wpautop for posts containing blocks

iseulde commented:
> I am very wary of the performance impact of detecting and individually formatting blocks in content during a front-end filter. [...] This is not complete compatibility, and it may be worth exploring whether wpautop should *also* be applied server-side when saving a post containing both blocks and freeform content to capture these instances.

What about just applying `autop` on freeform blocks server-side *before* save?

iseulde commented:
(This might capture two instances, and we wouldn't have to add a jus `autop` script.)

aduth commented:
One issue we might encounter is that TinyMCE will apply the paragraph tags to the freeform content, so on the initial edit of a legacy content post in Gutenberg, changes will be flagged even if they don't truly exist (merely focusing then blurring the freeform block). Since any post saved in Gutenberg would then have paragraphs applied (by a server-side filter in recent recommendations), I expect it'd only be an issue for that initial editing session.

This relates to Classic editor running `removep` on content before saving.

iseulde commented on blocks/api/parser.js:
```diff
@@ -165,8 +170,13 @@ export function getBlockAttributes( blockType, rawContent, attributes ) {
-	name = name || getUnknownTypeHandlerName();
+	// If unknown type, use unknown handler and apply paragraph formatting.
+	if ( undefined === name ) {
+		rawContent = formatting.autop( rawContent );
```
Hm, don't we already do that here: https://github.com/WordPress/gutenberg/blob/v1.1.0/blocks/library/freeform/old-editor.js#L79?

aduth replied:
>Hm, don't we already do that here: 

Good call, I did not see this, and it would likely be redundant. An issue with leaving the current implementation is that it needlessly flags the post as dirty because the original value of the block content does not have `autop` applied, only after setup has completed (in master, focusing and blurring a freeform block without changes will mark the post as needing save).

With ideas floating around server-side autop, I'll consider whether the existing `autop` call in `OldEditor` is still necessary.

aduth commented:
Reference to original approach, in anticipation of dropping commit during rebase: 392d5546e59e093679e2d8df81da419122893f3f

aduth committed:
Compat: Apply autop to freeform content for Gutenberg

aduth commented:
I've updated this branch with a revised approach, where the client is no longer expected at all to apply autop behavior. Instead, when loading Gutenberg, the content of the post is formatted to apply autop selectively to freeform content only (i.e. content is parsed and autop applied on a block-by-block basis).

To compare:

- We no longer need to maintain or account for differences in `wpautop` implementation differences between the client and server, since we rely entirely on the server
- We _could_ apply this on save, but due to the behavior described at https://github.com/WordPress/gutenberg/pull/2806#issuecomment-332522024, we would still want autop applied before the editor session begins, either on the server (as implemented currently) or in the client as part of the parse (difficult to faithfully recreate)
  - Even if another client would save with `removep`, if ever the user reopened the post in Gutenberg, the content should still be effectively the same

Where I hesitate:

- I don't like that we [modify the post entity](https://github.com/WordPress/gutenberg/pull/2806/files#diff-620130744b6b73b822a3df609671d930R733) from the server-side REST call when preparing the post for the editor. Unfortunately, there is already some precedent here in how we reset the Auto Draft title to an empty string. Alternatives here might be: Modifying content during REST filters, creating a separate content property (e.g. https://github.com/WordPress/gutenberg/pull/2649#issuecomment-326844673), or bootstrapping the page with a custom object shape for the post (not implying strict adherence to REST API schema).
- Recreating serialize behavior on the server seems useful, but I'm skeptical how well it mimics the behavior we have in the client: one in particular was slash encoding applied by default in [PHP's `json_encode`](http://php.net/manual/en/function.json-encode.php), which can be disabled in PHP 5.4+ with the `JSON_UNESCAPED_SLASHES` flag. Since we can't target 5.4+, I manually recreated this effect with a `str_replace`. I may need to check whether we need `JSON_UNESCAPED_UNICODE` here as well.

dmsnell commented on lib/blocks.php:
```diff
@@ -58,6 +58,72 @@ function gutenberg_parse_blocks( $content ) {
+		$attrs_json = json_encode( $block['attrs'] );
+
+		// In PHP 5.4+, we would pass the `JSON_UNESCAPED_SLASHES` option to
+		// `json_encode`. To support older versions, we must apply manually.
```
Nice comment!

Also, check out the JS serializer to see the other transforms we make in order to safeguard the attributes

https://github.com/WordPress/gutenberg/blob/e7f12238d84abab88afec915c1b94f7724bc88c8/blocks/api/serializer.js#L120

aduth replied:
>Also, check out the JS serializer to see the other transforms we make in order to safeguard the attributes

Ah, right, will make a pass to port over this behavior and equivalent tests.

dmsnell commented on lib/blocks.php:
```diff
@@ -93,3 +159,42 @@ function do_blocks( $content ) {
+		$blocks[ $i ]['rawContent'] = $content;
+	}
+
+	return gutenberg_serialize_blocks( $blocks );
```
This is one place where I have concern about using HTML strings as Editable (cc: @iseulde @youknowriad @mtias).

If we could iterate over the text nodes themselves and know that they contain no HTML then it seems like we could guarantee behaviors with `autop`. In this mixture, however, it feels like we leave the door open to bugs. Maybe it's not too big of a deal. I think `<` is illegal inside attributes and tags, but not illegal inside comments.

Not sure also how `autop` works so please read my review with that in mind.

aduth replied:
I'm not quite sure I understand the concern here. With these changes and #2708, we're effectively turning off `wpautop` for blocks, so it shouldn't be the case that we need to worry about how rich text (via HTML or other structure) is treated.

dmsnell replied:
more of a general concern about existing `apply_filter()` code

dmsnell commented on lib/register.php:
```diff
@@ -272,9 +272,21 @@ function gutenberg_can_edit_post( $post_id ) {
+/**
+ * Determine whether a content string contains blocks.
+ *
+ * @since 1.2.0
```
do you think it would be worth adding a note on the limited ability this has to accurately determine blocks? I would think it obvious from looking at the code but maybe someone wouldn't realize this is only a first-order approximation.

aduth replied:
Yeah, could be a good idea. The implementation is dubious, but... do you have in mind any scenarios in which it would not be accurate?

dmsnell replied:
just the rare case where someone writes something like…

```
This is not a <!-- wp --> block. <!-- wp:comment -->
```

…or when something odd like that accidentally crops up.

dmsnell commented on phpunit/fixtures/wpautop-original.html:
```diff
@@ -0,0 +1,23 @@
+
+Touch
+
+This
```
🔨 🕐

mkaz commented:
Testing this change and I don't get the initial fail state.

With a fresh check out and build from master, following the instructions above to create paragraph in Classic, and then in Gutenberg - both paragraphs get created with `<p>` tags around them. Editing also shows as expected with Classic Block and Gutenberg Paragraph Block.

I also tested on the branch with the change, and it behaves the same.

aduth committed:
Compat: Apply autop to freeform block content on post save

aduth commented:
Rebased to resolve conflicts, and with a few revisions included in the rebase.

- The rebased e8d802b2c1814efffff8c9970f7d1dec6f5aa33a ports string replacements for comment escaping from the JavaScript serializer to PHP, with an additional unit test for verifying serialized -> parsed -> serialized equivalence.
- The rebased 2d2775b40f41efe57787a09850f01152eba66b8b expands on the DocBlock to clarify accuracy of `gutenberg_content_has_blocks`. It could be reasonable to introduce a "strict" flag akin to the third argument of [PHP's `in_array`](http://php.net/manual/en/function.in-array.php) which would change the logic to running the parser and testing the result, but since we don't have an immediate need for this, it would be best to postpone for a future revision.
- I added 86a9792affe2ff70f49bf1db29eaea55064e91d3 which applies `wpautop` to freeform content for any save which includes block content. While this is arguably unnecessary and a waste of resources in parsing for saved content coming from the Gutenberg or classic editors, it could improve compatibility of saved content originating from other third-party editors which don't fall under Gutenberg or the classic editor changes from #2708 
- I omitted changes to remove the `gutenberg_` prefixing from block content functions. We may want to consider going down this path for easier merge, but should be done separately from this pull request.

dmsnell commented on lib/blocks.php:
```diff
@@ -58,6 +58,80 @@ function gutenberg_parse_blocks( $content ) {
+function gutenberg_serialize_block( $block ) {
+	// Return content of unknown block verbatim.
+	if ( ! isset( $block['blockName'] ) ) {
+		return $block['rawContent'];
```
quick ear-mark here as I think we're going to probably rename `rawContent` to something like `innerHtml`

aduth replied:
>quick ear-mark here as I think we're going to probably rename `rawContent` to something like `innerHtml`

I think that would be good. One idea I had originally was that, in parsing blocks, we could save effort in reserializing them since we should already know the original string text from which the block was parsed. Do you think it could be reasonable to still have something like `rawContent` returned from the parser which is assigned as the raw text of the block, comments and all?

Then again, there could be value in having the serializer available anyways, and not relying on blocks _maybe_ having this `rawContent` property available (e.g. not available if created from PHP, only if parsed from text).

dmsnell replied:
my original intention in bringing up `rawContent` was an _exception_ - that most blocks would return a tree of children but a few blocks which "need" the raw HTML would get it instead of the tree.

I can imagine use cases for the full body of the block but at the same time anticipate that providing as much is more of an enabler for working around the system than with it.

the way the PEG generator works this can be somewhat tricky to provide both - not sure why - which means that it may take more time if we want that to happen (we end up having to carry the `rawContent` attribute to each level of token and then reduce at the top level to stitch them together)

dmsnell reviewed (approved):
I did not test but I did review the code. Thanks for adding the comments on `has_blocky_stuff()`!

aduth commented:
Rebased to resolve conflicts. Would be good to have a few more eyes on this one.

mtias commented on lib/blocks.php:
```diff
@@ -58,6 +58,80 @@ function gutenberg_parse_blocks( $content ) {
 /**
+ * Given an array of parsed blocks, returns content string.
+ *
+ * @since 1.3.0
```
need to update these before merge

mkaz commented:
Testing out this change it works good, tested with various blocks of test with varying p tags and worked as expected.

However, the issue in https://github.com/WordPress/gutenberg/pull/2817 is back when Jetpack Markdown is enabled, I'm reopening that for @pento to take a look

aduth commented:
Noting that this branch incidentally resolves the issue where focusing then blurring a freeform block from a post authored in the classic editor would flag the post as needing to be saved.

aduth commented:
>However, the issue in https://github.com/WordPress/gutenberg/pull/2817 is back when Jetpack Markdown is enabled, I'm reopening that for @pento to take a look

By "back", do you mean reintroduced by these changes? Or is this issue present on master as well?

mkaz commented:
I see the issue still on master and confused by the state of #2817.
Primarily due to @pento [comment](https://github.com/WordPress/gutenberg/pull/2817#ref-commit-2bfed02) which was part of 2bfed02

> Revert the revert of #2817. 

The change he made looks right, so I think he was trying to put back my change that got overwritten.

mkaz commented:
ok, I'm pretty sure I know why it's not working, the loading of scripts changed from `gutenberg.php` to `lib/load.php` and the require for `lib/plugin-compat.php` was not included. 

I'll create a new PR for it.
